File: relation_extractor.py
"""
Relation Extraction Module
Extract relationships between entities using GPT-4o-mini
"""

import json
import time
from typing import List, Dict, Optional, Set
import spacy
from config import SPACY_MODEL
from openai_helper import OpenAIHelper


class RelationExtractor:
    """Extract relationships between entities with robust LLM calls"""
    
    def __init__(self, openai_api_key: Optional[str] = None):
        self.nlp = spacy.load(SPACY_MODEL)
        self.openai = OpenAIHelper(openai_api_key)
        self.use_llm = bool(openai_api_key)
    
    def extract_relations_pattern(self, text: str, entities: List[Dict]) -> List[Dict]:
        """Extract relations using dependency parsing patterns"""
        doc = self.nlp(text)
        relations = []
        
        entity_texts: Set[str] = {e['text'].lower() for e in entities}
        
        for sent in doc.sents:
            for token in sent:
                if token.dep_ in ('nsubj', 'nsubjpass'):
                    subject = token.text
                    verb = token.head.text
                    
                    for child in token.head.children:
                        if child.dep_ in ('dobj', 'pobj', 'attr'):
                            obj = child.text
                            
                            if subject.lower() in entity_texts and obj.lower() in entity_texts:
                                relations.append({
                                    'subject': subject,
                                    'predicate': verb,
                                    'object': obj,
                                    'confidence': 0.7
                                })
        
        print(f"✓ Extracted {len(relations)} relations using patterns")
        return relations
    
    def extract_relations_llm(self, text: str, entities: List[Dict], max_retries: int = 3) -> List[Dict]:
        """Extract relations using GPT-4o-mini with robust error handling
        
        Args:
            text: Input text to extract relations from
            entities: List of extracted entities
            max_retries: Number of retry attempts on failure
        """
        if not self.use_llm:
            print("⚠ LLM extraction skipped: No API key provided")
            return []
        
        if not entities:
            print("⚠ No entities provided for relation extraction")
            return []
        
        # Split text into chunks if too long (GPT-4o-mini has token limits)
        max_chunk_length = 6000  # Conservative limit for context
        if len(text) > max_chunk_length:
            return self._extract_relations_chunked(text, entities, max_retries)
        
        entity_list = [e['text'] for e in entities]
        
        # Enhanced prompt with clearer instructions and examples
        prompt = f"""You are an expert knowledge graph builder. Extract ALL meaningful relationships between the given entities from the text.

ENTITIES TO FIND RELATIONSHIPS FOR:
{', '.join(entity_list)}

TEXT:
{text}

INSTRUCTIONS:
1. Find relationships between ANY two entities from the list
2. Include various relationship types: worked_at, born_in, studied_at, discovered, developed, invented, founded, married_to, colleague_of, collaborated_with, published, awarded, etc.
3. Extract both explicit and implicit relationships
4. Use clear, consistent predicate names (lowercase, underscored)
5. Return ONLY valid JSON, no explanations

REQUIRED FORMAT (return as valid JSON array):
[
  {{"subject": "Entity1", "predicate": "relationship_type", "object": "Entity2"}},
  {{"subject": "Entity2", "predicate": "another_relationship", "object": "Entity3"}}
]

If no relationships found, return: []

JSON OUTPUT:"""
        
        messages = [
            {
                "role": "system", 
                "content": "You are an expert at knowledge graph construction and relationship extraction. You always return valid JSON arrays."
            },
            {
                "role": "user", 
                "content": prompt
            }
        ]
        
        # Retry logic with exponential backoff
        for attempt in range(max_retries):
            try:
                print(f"  Attempting LLM relation extraction (attempt {attempt + 1}/{max_retries})...")
                
                # Use GPT-4o-mini for faster and cheaper extraction
                response = self.openai.chat_completion(
                    messages, 
                    temperature=0.3,  # Lower temperature for more consistent output
                    model="gpt-4o-mini"
                )
                
                if not response:
                    print(f"  ✗ Attempt {attempt + 1}: Empty response from API")
                    if attempt < max_retries - 1:
                        time.sleep(2 ** attempt)  # Exponential backoff
                        continue
                    return []
                
                # Parse the response
                relations = self._parse_llm_response(response)
                
                if relations is not None:
                    # Validate relations
                    valid_relations = self._validate_relations(relations, entities)
                    print(f"✓ Extracted {len(valid_relations)} relations using LLM (GPT-4o-mini)")
                    return valid_relations
                else:
                    print(f"  ✗ Attempt {attempt + 1}: Failed to parse JSON response")
                    if attempt < max_retries - 1:
                        time.sleep(2 ** attempt)
                        continue
                    
            except Exception as e:
                print(f"  ✗ Attempt {attempt + 1}: LLM extraction error: {e}")
                if attempt < max_retries - 1:
                    time.sleep(2 ** attempt)
                    continue
        
        print("✗ LLM relation extraction failed after all retries")
        return []
    
    def _extract_relations_chunked(self, text: str, entities: List[Dict], max_retries: int) -> List[Dict]:
        """Extract relations from long text by splitting into chunks"""
        print("  Text too long, splitting into chunks...")
        
        # Split into sentences
        doc = self.nlp(text)
        sentences = [sent.text for sent in doc.sents]
        
        # Group sentences into chunks
        chunks = []
        current_chunk = []
        current_length = 0
        max_chunk_length = 5000
        
        for sent in sentences:
            if current_length + len(sent) > max_chunk_length and current_chunk:
                chunks.append(" ".join(current_chunk))
                current_chunk = [sent]
                current_length = len(sent)
            else:
                current_chunk.append(sent)
                current_length += len(sent)
        
        if current_chunk:
            chunks.append(" ".join(current_chunk))
        
        print(f"  Processing {len(chunks)} chunks...")
        
        all_relations = []
        for i, chunk in enumerate(chunks):
            print(f"  Chunk {i+1}/{len(chunks)}...")
            chunk_relations = self.extract_relations_llm(chunk, entities, max_retries=1)
            all_relations.extend(chunk_relations)
        
        print(f"✓ Extracted {len(all_relations)} total relations from all chunks")
        return all_relations
    
    def _parse_llm_response(self, response: str) -> Optional[List[Dict]]:
        """Parse LLM response and extract JSON"""
        try:
            content = response.strip()
            
            # Remove markdown code blocks
            if "```json" in content:
                content = content.split("```json")[1].split("```")[0]
            elif "```" in content:
                content = content.split("```")[1].split("```")[0]
            
            # Clean up common issues
            content = content.strip()
            
            # Try to parse JSON
            relations = json.loads(content)
            
            # Ensure it's a list
            if not isinstance(relations, list):
                print(f"  ✗ Response is not a list: {type(relations)}")
                return None
            
            return relations
            
        except json.JSONDecodeError as e:
            print(f"  ✗ JSON parsing error: {e}")
            print(f"  Response preview: {response[:200]}...")
            return None
        except Exception as e:
            print(f"  ✗ Unexpected parsing error: {e}")
            return None
    
    def _validate_relations(self, relations: List[Dict], entities: List[Dict]) -> List[Dict]:
        """Validate and clean extracted relations"""
        valid_relations = []
        entity_texts = {e['text'].lower() for e in entities}
        
        for rel in relations:
            # Check required fields
            if not all(key in rel for key in ['subject', 'predicate', 'object']):
                continue
            
            # Check if subject and object are valid
            subject = str(rel['subject']).strip()
            predicate = str(rel['predicate']).strip()
            obj = str(rel['object']).strip()
            
            if not subject or not predicate or not obj:
                continue
            
            # Normalize predicate (lowercase, replace spaces with underscores)
            predicate = predicate.lower().replace(' ', '_')
            
            valid_relations.append({
                'subject': subject,
                'predicate': predicate,
                'object': obj,
                'confidence': 0.9  # Higher confidence for LLM extractions
            })
        
        return valid_relations
    
    def merge_relations(self, pattern_relations: List[Dict], llm_relations: List[Dict]) -> List[Dict]:
        """Merge and deduplicate relations, prioritizing LLM results"""
        relation_map = {}
        
        # Add pattern-based relations first
        for rel in pattern_relations:
            key = (rel['subject'].lower(), rel['predicate'].lower(), rel['object'].lower())
            relation_map[key] = rel
        
        # Override/add with LLM relations (higher quality)
        for rel in llm_relations:
            key = (rel['subject'].lower(), rel['predicate'].lower(), rel['object'].lower())
            relation_map[key] = rel
        
        merged = list(relation_map.values())
        print(f"✓ Merged into {len(merged)} unique relations")
        
        if merged:
            print(f"  Sample relations: {merged[:3]}")
        
        return merged

File: backend/src/main/kotlin/io/github/juantoanxoup/kg/RelationExtractor.kt
package io.github.juantoanxoup.kg

import ai.koog.prompt.dsl.prompt
import ai.koog.prompt.executor.model.PromptExecutor
import ai.koog.prompt.executor.model.StructureFixingParser
import ai.koog.prompt.executor.model.executeStructured
import ai.koog.prompt.params.LLMParams
import edu.stanford.nlp.pipeline.CoreDocument
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import org.slf4j.LoggerFactory

/** A (subject, predicate, object) relation between two entity surface strings. */
@Serializable
data class Relation(
    val subject: String,
    val predicate: String,
    @SerialName("object") val obj: String,
    val confidence: Double,
)

/** Structured output requested from the LLM. */
@Serializable
data class ExtractedRelations(
    val relations: List<ExtractedRelation>,
)

@Serializable
data class ExtractedRelation(
    val subject: String? = null,
    val predicate: String? = null,
    @SerialName("object") val obj: String? = null,
)

/** Relation extraction with dependency patterns and an LLM (relation_extractor.py). */
class RelationExtractor(
    private val executor: PromptExecutor? = OpenAiHelper.promptExecutor(),
) {
    private val log = LoggerFactory.getLogger(RelationExtractor::class.java)

    /**
     * Pattern pass over Universal Dependencies: `nsubj`/`nsubj:pass` gives subject and verb,
     * `obj`/`iobj`/`obl*` children of the verb give the object. Kept only when both ends are known entities.
     */
    fun extractRelationsPattern(
        text: String,
        entities: List<Entity>,
    ): List<Relation> {
        val doc = CoreDocument(text)
        Nlp.pipeline.annotate(doc)
        val entityTexts = entities.map { it.text.lowercase() }.toSet()
        val relations = mutableListOf<Relation>()

        for (sentence in doc.sentences()) {
            val graph = sentence.dependencyParse() ?: continue
            for (edge in graph.edgeIterable()) {
                val relation = edge.relation.toString()
                if (relation != "nsubj" && relation != "nsubj:pass") continue
                val subject = edge.dependent.word()
                val verb = edge.governor.word()
                for (child in graph.outgoingEdgeIterable(edge.governor)) {
                    val childRelation = child.relation.toString()
                    if (childRelation == "obj" || childRelation == "iobj" || childRelation.startsWith("obl")) {
                        val obj = child.dependent.word()
                        if (subject.lowercase() in entityTexts && obj.lowercase() in entityTexts) {
                            relations += Relation(subject, verb, obj, PATTERN_CONFIDENCE)
                        }
                    }
                }
            }
        }
        log.info("Extracted {} relations using patterns", relations.size)
        return relations
    }

    /**
     * LLM pass as a structured-output request. Transport retries come from Koog's retrying client and
     * malformed JSON is repaired by [StructureFixingParser]; long texts are chunked by sentence.
     */
    suspend fun extractRelationsLlm(
        text: String,
        entities: List<Entity>,
    ): List<Relation> {
        val llm = executor
        if (llm == null) {
            log.warn("LLM extraction skipped: No API key provided")
            return emptyList()
        }
        if (entities.isEmpty()) {
            log.warn("No entities provided for relation extraction")
            return emptyList()
        }
        if (text.length > MAX_TEXT_LENGTH) return extractRelationsChunked(text, entities)

        val entityList = entities.joinToString(", ") { it.text }
        val request =
            prompt("relation-extraction", LLMParams(temperature = LLM_TEMPERATURE)) {
                system(
                    "You are an expert at knowledge graph construction and relationship extraction. You always return valid JSON.",
                )
                user(
                    """
                    You are an expert knowledge graph builder. Extract ALL meaningful relationships between the given entities from the text.

                    ENTITIES TO FIND RELATIONSHIPS FOR:
                    $entityList

                    TEXT:
                    $text

                    INSTRUCTIONS:
                    1. Find relationships between ANY two entities from the list
                    2. Include various relationship types: worked_at, born_in, studied_at, discovered, developed, invented, founded, married_to, colleague_of, collaborated_with, published, awarded, etc.
                    3. Extract both explicit and implicit relationships
                    4. Use clear, consistent predicate names (lowercase, underscored)
                    5. If no relationships are found, return an empty list
                    """.trimIndent(),
                )
            }
        val result =
            llm.executeStructured<ExtractedRelations>(
                prompt = request,
                model = OpenAiHelper.defaultModel,
                fixingParser = StructureFixingParser(OpenAiHelper.defaultModel, retries = FIXING_RETRIES),
            )
        return result
            .map { validateRelations(it.data.relations) }
            .onSuccess { log.info("Extracted {} relations using LLM ({})", it.size, OpenAiHelper.defaultModel.id) }
            .onFailure { log.error("LLM relation extraction failed: {}", it.message) }
            .getOrDefault(emptyList())
    }

    private suspend fun extractRelationsChunked(
        text: String,
        entities: List<Entity>,
    ): List<Relation> {
        log.info("Text too long, splitting into chunks...")
        val doc = CoreDocument(text)
        Nlp.sentenceSplitter.annotate(doc)
        val chunks = chunkSentences(doc.sentences().map { it.text() }, CHUNK_LENGTH)
        log.info("Processing {} chunks...", chunks.size)

        val all = mutableListOf<Relation>()
        chunks.forEachIndexed { i, chunk ->
            log.info("Chunk {}/{}...", i + 1, chunks.size)
            all += extractRelationsLlm(chunk, entities)
        }
        log.info("Extracted {} total relations from all chunks", all.size)
        return all
    }

    /** Groups sentences into chunks of at most [maxChunkLength] characters. */
    internal fun chunkSentences(
        sentences: List<String>,
        maxChunkLength: Int,
    ): List<String> {
        val chunks = mutableListOf<String>()
        val current = mutableListOf<String>()
        var currentLength = 0
        for (sentence in sentences) {
            if (currentLength + sentence.length > maxChunkLength && current.isNotEmpty()) {
                chunks += current.joinToString(" ")
                current.clear()
                current += sentence
                currentLength = sentence.length
            } else {
                current += sentence
                currentLength += sentence.length
            }
        }
        if (current.isNotEmpty()) chunks += current.joinToString(" ")
        return chunks
    }

    /** Requires non-empty subject, predicate, object; normalizes the predicate; assigns LLM confidence. */
    internal fun validateRelations(relations: List<ExtractedRelation>): List<Relation> =
        relations.mapNotNull { rel ->
            val subject = rel.subject?.trim().orEmpty()
            val predicate = rel.predicate?.trim().orEmpty()
            val obj = rel.obj?.trim().orEmpty()
            if (subject.isEmpty() || predicate.isEmpty() || obj.isEmpty()) return@mapNotNull null
            Relation(subject, predicate.lowercase().replace(' ', '_'), obj, LLM_CONFIDENCE)
        }

    /** Keyed by lower-cased triple; LLM relations overwrite pattern relations on collision. */
    fun mergeRelations(
        patternRelations: List<Relation>,
        llmRelations: List<Relation>,
    ): List<Relation> {
        val map = LinkedHashMap<Triple<String, String, String>, Relation>()
        for (rel in patternRelations) map[rel.key()] = rel
        for (rel in llmRelations) map[rel.key()] = rel
        val merged = map.values.toList()
        log.info("Merged into {} unique relations", merged.size)
        if (merged.isNotEmpty()) log.info("Sample relations: {}", merged.take(3))
        return merged
    }

    private fun Relation.key() = Triple(subject.lowercase(), predicate.lowercase(), obj.lowercase())

    private companion object {
        const val PATTERN_CONFIDENCE = 0.7
        const val LLM_CONFIDENCE = 0.9
        const val LLM_TEMPERATURE = 0.3
        const val MAX_TEXT_LENGTH = 6000
        const val CHUNK_LENGTH = 5000
        const val FIXING_RETRIES = 2
    }
}
